### DISCUSSION TITLE
Welcome to ACadSharp Discussions!

### DISCUSSION THREAD

**Discussion by DomCR** [DomCR/ACadSharp#89] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ottosson** [comment]
Hi! I'm building an application that utilizes AutoCAD as the engine to read/write dwg files. This is not optimal as you need to have AutoCAD installed and pay for a license to be able to use the application. So I would very much like to see this project succeed! Anything that takes away the DWG monopoly from Autodesk is a good thing imo.

How have you learned about the DWG-format? Are you looking at the [Open Design Specification for .dwg files](https://www.opendesign.com/files/guestdownloads/OpenDesign_Specification_for_.dwg_files.pdf) or where did you learn the inner workings of the file format?

Thanks!

**Reply by DomCR** [reply]
Hi @ottosson,

This library is mainly created based on the Open Design Specification for dwg, if you check the code you will see the different steps of the document as comments in the code and for the classes and objects of the file I've used the Dxf format specification for each one of them.

**Reply by ottosson** [reply]
Ok, thanks! As someone who's read and understands the specification (I've only scratched the surface), would you say it has everything (or most) of the format covered? Is something missing?

Autodesk haven't released a new dwg-version in a couple of years now and looking at the history it's not unlikely they'll release a new version soon. What do you think happens when they do? Will they release an update to the specification or will it be up to the community to reverse engineer the format, if that is even remotely feasible? I know this is more or less impossible to know, I'm just curious what you think :)

**Reply by DomCR** [reply]
The document covers all parts of the DWG format, in some specific cases the documentation does not know what a value is or what it represents and works with the defaults, in some cases I've managed to know the different values by manually reading the values from a DXF ASCII file.

About the new version, I'm sure that Autodesk will release some new versions and to be honest I'll address that problem when whenever it comes, by now the library is starting to work and give good results when reading the current versions and Autocad always allow to save in older versions to guarantee backwards compatibility so I'm not worried about that at the moment.

**Reply by ottosson** [reply]
That's good to know. I wish you good luck with this project. I will follow it closely, and if I feel there's something I can contribute I will.
Thanks!

**Comment by thorster** [comment]
I am writing a program that creates Excel document "briefcases" that contain a lists of documents.  These documents are in xlsx or dwg-format, Text from the document headers is extracted and written to to the briefcase which contains one row per document.

I have a first version of my program ready and AcadSharp works very well. I came across two bugs in AcadSharp that prevented some dwg-files from being read. I reported the bugs and within a day @DomCR had fixed the bugs and commited to github. I am impressed by the package, it is easy to use and performance is great. Thank you @DomCR for sharing AcadSharp and your commitment to solve issues,

I have previously developed a .NET-plugin that peforms various batch processing on dwg-files. It requires an installation of either AutoCAD or BricsCAD. I will later look into if AcadSharp can be an alternative. 

Developing is a hobby of mine and I don't think I have enough skills to contribute to the code base unless it something not so advanced. If you need help testing I migth be able to assist. I have access to large amounts of dwg-files.

**Reply by DomCR** [reply]
Hi @thorster,

The community using the library is one of the things that makes the project grow and improve. Is really difficult to manage all the possible edge cases for the different forms of the Dwg/Dxf files, if you are working with a file that causes problems when is opened by `ACadSharp` please open an issue and I'll look into it, this is the kind of collaboration that helps the project because adds new test cases and helps fixing bugs that otherwise I cannot find.

Thanks a lot for your support in this project!

**Comment by douglaseds9** [comment]
To develop software that writes and reads dwg files, would it be necessary to pay royalties or some license to Autodesk?

**Reply by DomCR** [reply]
Hi @douglaseds9,

This library is entirely based on the document publish by [OpenDesign](https://www.opendesign.com/files/guestdownloads/OpenDesign_Specification_for_.dwg_files.pdf) which is public for everyone so anyone could develop a reader/writer for the DWG format.

There are other open source projects that also write and read DWG files like [LibreDWG](https://github.com/LibreCAD/LibreCAD), FreeCAD or QCAD among others. 

I hope that I've answered your question.
